### PR TITLE
fix #1994 gorge, staticExec now raise Assert on exitCode !=0; simplified implementation

### LINKED ISSUE #1994: Can't tell return value of programs with staticExec
Using `staticExec` the return value of the executed program can't be checked. Even the manual's example of `git rev-parse HEAD` can fail if you're not inside a git repo, which you can't catch nicely. I think `staticExec` should either return an empty string if the return value isn't 0, or have a way to get back the return value as well as stdout.

### PR DESCRIPTION
**Pull request by timotheecour**
* fix #1994

### CODE REVIEW

File: changelog.md [hunk 1]
```diff
@@ -67,7 +67,9 @@
 - two poorly documented and not used modules (`subexes`, `scgi`) were moved to
   graveyard (they are available as Nimble packages)
 
-
+- `gorge`, `staticExec` now raise AssertError if exitCode is not 0 instead of
```
Comment (dom96) [inline changelog.md]: Assertions aren't catchable, I think we want something that is catchable to be raised.

File: tests/vm/tvmops.nim [hunk 2]
```diff
@@ -5,6 +5,8 @@ import os
 import math
 import strutils
 
+const unexistant = "D20190116T211842"
```
[the remark this answers is not in the document]
Comment (krux02) [reply tests/vm/tvmops.nim]: Why did you introduce this constant?

**Comment by krux02** [comment]
one possible solution is like in  bash, it the exit code always in a special variable, but that is bad in multithreaded environments.

You should rename the title of this review. With the title how it is now I wanted to close this PR without even looking at the code, because an assert would break my usages of staticExec.

File: tests/vm/tvmops.nim (hunk 2 above)
[the remark this answers is not in the document]
Comment (timotheecour) [reply tests/vm/tvmops.nim]: > Minor nitpick: s/unexistant/nonexistent/

done

> Why did you introduce this constant?

to make it unlikely it ever exists whatever the conditions; it's more unlikely than "nonexistant"

File: changelog.md (hunk 1 above)
Comment (timotheecour) [reply changelog.md]: but that's at compile time; user can always use the cleaner `gorgeEx` variant which returns exitCode and handle accordingly so there's not loss of usability.

In any case, catching doesn't work:
```
static:
  try:
    let a = gorge("D20190116T211842")
  except OSError as e:
    echo "caught"
```

gives:
```
Error: cannot evaluate at compile time: currException
```

**Comment by timotheecour** [comment]
> You should not just break the compatibility with existing code that uses the existing API. Please find a deprecation path that doesn't break existing code.

discussed in https://github.com/nim-lang/Nim/pull/10360

> one possible solution is like in bash, it the exit code always in a special variable, but that is bad in multithreaded environments.

that's too error prone

> You should rename the title of this review. With the title how it is now I wanted to close this PR without even looking at the code, because an assert would break my usages of staticExec.

a PR shouldn't be closed just because there's 1 thing you dislike about it in the title, especially when that thing can be addressed during review. 

* as I've explained above, if I call `raise newException(OSError)` instead of calling `doAssert`, attempt at catching it results in `Error: cannot evaluate at compile time: currException` (see https://github.com/nim-lang/Nim/pull/10345#discussion_r248960535)

* I'm hitting a technical wall with this approach, which is the test failure for `./tests/vm/tgorge.nim
`gorge` has the odd (and undocumented) behavior of setting `workingDir` to `foo` if `gorge` is called from `foo/bar.nim`; I couldn't figure out how to keep that behavior with approach in this PR that makes `gorge` a regular proc calling `gorgeEx`, because: if `gorge` is a regular proc, we have no knowledge (at CT) of caller; if `gorge` is made a template (and using `instantiationInfo`), it also leads to other issues that I cannot solve (without changing the compiler).

because of that, I'm closing this in favor of https://github.com/nim-lang/Nim/pull/10360 but if someone has a concrete idea (and has tested it to work!) for how to make it work, please let me know and I'll reopen this PR as it's cleaner in some ways

**Pull request closed without merging**